---
title: "High-Protein BigCages - mapping statistics"
author: "Claudia Ramirez-Lanzas"
date: "02/05/2023"
---
  
Part1: Mapping raw data to the reference genome.
Part2: SubRead count table statistics

Note that replicates from R31 to R36 correspond to replicates R1 to R6 in the text.

# Load libraries:
```{r}
# Data handling:
library("readxl")
library("writexl")
library("xlsx")
library("tidyverse")

```

# Import files: 
```{r}
# Import the mapping statistics generated by SAMtools:
mapping_stats <- read_csv("input_files/BigCages_mapping_stats.csv", show_col_types = FALSE)

# Import the statistics generated by multiQC on the SubRead summary:
subread_stats <- read_csv("input_files/BigCages_multiqc_featureCounts.csv", show_col_types = FALSE)

```

*mapping_stats: .csv containing the samtools statistics output (SN section only) of the mapping by the gmap software.
*subread_stats: .csv containing the multiQC statistics of the featureCounts output by the subRead software.

The sample_seq_id contains the biological name of each sample in both data-sets.

# Part1: Mapping raw data to the reference genome
# Plot - total reads
```{r}
# total reads.
ggplot(data=mapping_stats, aes(x=sample_description, y=sequences)) + 
  geom_col() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) + 
  labs(title="Total sequences")

```
# Plot - Fraction of mapped reads in relation to total reads.
```{r}
ggplot(data=mapping_stats, aes(x=sample_description, y=mapped_fraction)) + 
  geom_col() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) + 
  ylim(0, 1) + labs(title="Fraction of mapped reads")

```
# Plot - Fraction of mapped and paired reads in relation to all reads.
```{r}
ggplot(data=mapping_stats, aes(x=sample_description, y=mapped_and_paired_fraction)) + 
  geom_col() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) +
  ylim(0, 1) + labs(title="Fraction of paired and mapped reads")

```
# Plot - Average read quality.
```{r}
ggplot(data=mapping_stats, aes(x=sample_description, y=`average quality:`)) + 
  geom_col() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) + 
  labs(title="Average mapping quality")

```
# Plot - Average insert size 
We expect an average insert size close to 300bp
```{r}
ggplot(data=mapping_stats, aes(x=sample_description, y=`insert size average:`)) + 
  geom_col() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) + 
  labs(title="Insert size average")

```
# Plot - Pair reads mapping to different chromosomes
```{r}
ggplot(data=mapping_stats, aes(x=sample_description, y=`pairs on different chromosomes:`)) + 
  geom_col() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) + 
  labs(title="pair reads mapping to different chromosomes") +
  ylab("pairs on different chromosomes")

```
# Plot - Total raw reads, mapped reads and maaped and paired reads:
```{r}
# total raw, mapped and mapped-paired reads.
ggplot(data = mapping_stats, aes(x = sample_description)) + 
  geom_col(aes(y = sequences, fill = "sequences"), position = "dodge", width = 0.4) +
  geom_col(aes(y =`reads mapped:`, fill = "reads_mapped"), position = "dodge", width = 0.4) +
  geom_col(aes(y = `reads mapped and paired:`, fill = "reads_mapped_and_paired"), position = "dodge", width = 0.4) +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) + 
  labs(title = "Total raw reads, mapped reads and mapped-paired reads",
       y = "Number of reads",
       fill = "") +
  scale_fill_manual(values = c("sequences" = "green", "reads_mapped" = "red", "reads_mapped_and_paired" = "black"))

```
# Part2: SubRead count table statistics
# Plot - paired fragments Mapped to the exome
```{r}
subread_stats %>% ggplot(aes(x = sample_description , y = Assigned)) +
  geom_bar(stat = "identity") + ylab("Total paired fragments") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

--- END ---



