# BigCages_high_protein

This folder contains all the scripts needed to reproduce the analyses in:
**Tracking polygenic adaptation: phenotypic stasis after rapid adaptation in large populations of *Drosophila* ** 
Claudia Ramirez-Lanzas, Marlies Dolezal, Sarah Signor, Christian Schlötterer, and Neda Barghi  

## 01_Fecundity.Rmd
This R notebook contains the pipeline to plot fecundity and perform statistical analyses.

Input 1: BigCages_fecundity_weight_per_transfer.csv\

Output 1: BigCages_fecundity_total.csv\ 

## 02_Dry_weight.Rmd
This R notebook contains the pipeline to plot dry weight and perform statistical analyses.

Input 1: BigCages_dry_weight.csv\ 

## 03_TAG_quantification.Rmd
This R notebook contains the pipeline to compute TAG concentrations from
absorbance values and using standard curves.

Input 1: BigCages_TAG_540nm_absorbances.csv\
Input 2: BigCages_TAG_540nm_standards.csv\

## 04_TAG.Rmd
This R notebook contains the pipeline to plot TAG content and perform statistical analyses.

Input 1: BigCagesTAG_concentrations.csv\

## 05_RNAseq_mapping_stats.Rmd
This R notebook contains the code to plot the mapping statistics of RNAseq
samples to the reference genome using gmap software as well as the statistics of
the count table generated by subRead software (featureCounts function).

Input 1: BigCages_mapping_stats.csv\
Input 2: BigCages_multiqc_featureCounts.csv\

## 06_RNA_PCA.Rmd

This R notebook contains the code to perform and visualize the
Principal Component Analysis of the transcriptomic data.

Input 1: count_table.csv\

## 07_DE_variancePartition.Rmd

This R notebook contains the pipeline followed to model time-series gene
expression and to perform differential expression (DE) analysis. 

- Modeling overall response
- DE overall response
- Modeling replicate-specific response
- DE overall replicate-specific response

Input 1: count_table.csv\
  
Output 1: DE_overall_results.csv  overall response DE results: log2FC, adjusted p-values etc.\
Output 2: DE_repwise_results.csv replicate-specific response DE results: log2FC, adjusted p-values etc.\

## 08_Transcriptomic_trajetories.Rmd

This R notebook contains the pipeline followed to classify the gene expression
trajectories across generations.   

- Quantification of plateau gene expression
- Plotting trajectories

Input 1: DE_overall_results.csv\
Input 2: DE_repwise_results.csv\
Input 3: count_table.csv\

-- Output 1: classify_trajectories_df.csv DE results classified by trajectories.\

##  09_ Compute_gene_expression_variance.Rmd
This R notebook contains the code to compute gene variance using log2CPM expression
values.

Input 1: classify_trajectories_df.csv\  
Input 1: count_table.csv\

Output 1: variance_per_gene.csv dataset with the gene variances.\
Output 2: S3_gene_variance_across_groups_tests.xlsx dataset with the statistic
results of group contrasts\
  
## 10_Transcriptomic_parallelism.Rmd

This R notebook contains the pipeline followed to compute Jaccard Index and
Pearson´s correlation on transcriptomic responses across replicates as
measures of parallelism.   

- Compute pairwise Jaccard Index across 6 replicates for each generation.
- Compute Pearson´s correlation using log2FC. 

Input 1: classify_trajectories_df.csv\ 
Input 2: DE_repwise_results.csv\

## 11_Compute_Tau_from_flyatlas2.Rmd
This R notebook contains the pipeline followed to compute tissue specificity
for each gene based on their expression levels across tissues. 

Input 1: FlyAtlas2_gene_data_2023.csv tissue specific gene expression dataset
downloaded from https://motif.mvls.gla.ac.uk/FlyAtlas2 (downloaded on September 2023).\

Output 1: au_flyatlas2_FKM_male.csv dataset containing tau values for each gene.\

## 12_Pleiotropy.Rmd
This R notebook contains the pipeline followed to quantify pleiotropy for each
gene using two metrics (connectivity and tissue specificity).   

- Connectivity
- Tissue specificity

Input 1: tau_flyatlas2_FKM_male.csv\
Input 2: flynet_supervised_0.6.txt -> D. mel regulatory network:
contains ~300,000 edges of a supervised integrative gene
regulatory network of Drosophila melanogaster from
(Marbach et. al., Genome Research, 2012).\
Input 3: classify_trajectories_df.csv\

Output 1: connectivity_per_gene.csv\
Output 2: oneminusTau_per_gene.csv (1-Tau) per gene\
Output 3: S5_wilcox_test_pleiotropy.xlsx pleiotropy levels across transcriptomic groups tests\
Output 4: S6_correlation_pleiotropy_log2FC Kendall´s correlation coefficients and tests \

## 13 Relationship_pleiotropy_and_parallelism.Rmd
This R notebook contains the pipeline followed to plot and test the relationship
between levels of pleiotropy and transcriptomic parallelism.

Input 1: connectivity_per_gene.csv\
Input 2: tau_per_gene.csv\
Input 3: shared_upset_data.csv\
Input 4: DE_repwise_results.csv\

Output 1: S7_wilcox_test_pleiotropy_and_parallellism.xlsx pleiotropy levels
across parallelism groups tests.\


## 14_Enrichment_analysis.Rmd

This R notebook contains the pipeline followed to perform enrichment tests on
specific groups of genes. The tests are:

- GO enrichment using TopGo library
- Tissue enrichment using flyAtlas2 database
- Pathway enrichment using KGEE database

Input 1: classify_trajectories_df.csv\
Input 2: DE_overall_results.csv\
Input 3: FlyAtlas2_gene_data_2023.csv\
Input 4: DE_repwise_results.csv\
Input 5: shared_upset_data.csv\
 
Output 1: GO_enrichment_results_BP.xlsx\
Output 2: tissue_enrichment_results.xlsx\
Output 3: KEGG_enrichment_results.xlsx\

## 15_Darwin_numerator_HighOrder_pheno.Rmd

This R notebook contains the pipeline followed to compute Darwin numerators across
high order phenotypes of our study and other studies shown in Fig. 4 A.

Input 1: BigCages_fecundity_total.csv (our study)\
Input 2: BigCages_dry_weight.csv (our study)\
Input 3: BigCagesTAG_concentrations.csv (our study)\
Input 4: Rudman_OutdoorCages_AdaptTrack_FecundData_211222.csv\
Input 5: Barghi_LNS_HotI_fecundity.csv\ 
Input 6: Barghi_LNS_HotI_HotbodyFatContent.csv\
Input 7: Barghi_LNSI_metabolism.csv\
Input 8: Mallard_LNS_HotII_Cold_fecundity.csv\
Input 9: Mallard_LNS_HotII_Cold_fecundity_counts.csv\

Output 1: Darwin_numerators_all_studies_HighOrder_pheno.csv contains the Darwin
numerators as well as Haldanes and Darwins for all traits.\ 

## 16_Darwin_numerator_gene_expression.Rmd

This R notebook contains the pipeline followed to compute and compare
Darwin numerators for gene expression in our study and other studies shown
in Fig. 4 B and C.

Input 1: count_table.csv (our study)\
Input 2: classify_trajectories_df.csv (our study)\
Input 3: DE_overall_results.csv (our study)\
Input 4: Souto_Maior_Artificial_Selection_GSE202600_Normalized_read_counts.csv\
Input 5: Hsu_LNS_HotI_RNASeq_CGE1_count_table_transcriptome.csv\
Input 6: Mallard_LNS_HotII_Cold_Final_counts_Dryad_transcriptome.csv\

## 17_Transcriptomic_trajectories_after_filter_by_log2FC.Rmd

This R notebook contains the pipeline where the transcriptomic trajectories are
quantified after applying a filtering by log2FC removing all genes with
log2FC < 0.32  (~0.25 fold change).

Input 1: DE_overall_results.csv
Input 2: count_table.csv

## 18_Artificial_selection_transcriptomic_analysis.Rmd
This R notebook contains the pipeline to perform principal component analysis (PCA),
differential expression analysis and transcriptomic trajectories quantification
in order to see the transcriptomic patterns of populations under truncating selection
(directional selection) and compare with our study populations that are close to
the equilibrium or stasis phase.

Here we use the transcriptomic data published by
C. Souto-Maior et al., PLOS Comput. Biol (2023)

-- Input 1: Souto_Maior_Artificial_Selection_GSE202600_Normalized_read_counts.csv\


## 19_Diet_composition.Rmd
Code to plot the composition of Standard and High Protein diets in Fig. S3.

--- END ---